packages/lodestar/src/network/attestationService.ts

```diff
@@ -343,30 +337,38 @@ export class AttestationService implements IAttestationService {
    * allocated amount of random subnets.
    */
   private handleKnownValidatorExpiry(currentSlot: Slot): void {
-    const expiredValidators: Set<ValidatorIndex> = new Set();
     for (const [index, slot] of this.knownValidators.entries()) {
-      if (currentSlot > slot) expiredValidators.add(index);
-    }
-    const numConnectedValidators = this.knownValidators.size - expiredValidators.size;
-    if (numConnectedValidators * this.config.params.RANDOM_SUBNETS_PER_VALIDATOR >= ATTESTATION_SUBNET_COUNT) {
-      // have too many validators
-      for (const validator of expiredValidators) {
-        this.knownValidators.delete(validator);
-      }
-      return;
+      if (currentSlot > slot) this.knownValidators.delete(index);
     }
+    this.rebalanceRandomSubnets();
+  }
 
+  /**
+   * Called when we have new validators or expired validators.
+   * knownValidators should be updated before this function.
+   */
+  private rebalanceRandomSubnets(): void {
+    const {RANDOM_SUBNETS_PER_VALIDATOR} = this.config.params;
+    const numValidators = this.knownValidators.size;
+    // random subnet is renewed automatically
+    let numRandomSubnets = this.randomSubnets.getAll().length;
+    // subscribe to more random subnets
+    const currentFork = this.chain.getForkName();
+    const targetRandomSubnetCount = numValidators * RANDOM_SUBNETS_PER_VALIDATOR;
+    while (targetRandomSubnetCount > numRandomSubnets && numRandomSubnets < ATTESTATION_SUBNET_COUNT) {
+      this.subscribeToRandomSubnet(currentFork);
```

Comment:
You pass the number of new subnets to subscribe to here and won't need to iterate with a while. The logic inside `subscribeToRandomSubnet` can be re-used if subscribing to multiple subnets at once. Consider moving the `ATTESTATION_SUBNET_COUNT` check there too to compact the logic

Reply:
Yeah, might be a small gain, but agree, `subscribeToRandomSubnet` could take a second `count` param. Then we only filter/select subnets / update the ENR once.